src/java/org/apache/cassandra/index/sai/disk/vector/OnDiskOrdinalsMap.java

```diff
@@ -231,17 +233,42 @@ private class FileReadingOrdinalsView implements OrdinalsView
     {
         RandomAccessReader reader = fh.createReader();
         private final long high = (segmentEnd - 8 - rowOrdinalOffset) / 8;
+        private final AtomicInteger lastFoundRowId = new AtomicInteger(-1);
+        private final AtomicLong lastFoundRowIdIndex = new AtomicLong(-1);
 
         /**
          * @return order if given row id is found; otherwise return -1
          */
         @Override
         public int getOrdinalForRowId(int rowId) throws IOException
         {
+            long low = 0;
+            if (lastFoundRowId.get() > -1 && lastFoundRowId.get() <= rowId)
```

Comment:
All current access patterns monotonically increase `rowId`. I think we can take advantage of that here by:

1. asserting that the `rowId` is greater than the `lastRowId`
2. returning `-1` if `rowId < lastFoundRowId` 
3. if `rowId == lastFoundRowId`, seek to `lastFoundRowIdIndex` and read the ordinal

Reply:
(2) results in non-deterministic function, I'd avoid this.
(3) is fine, but then it means we anticipate changes in read patterns, so I'll skip assert and log warning instead.

Reply:
> (2) results in non-deterministic function, I'd avoid this.

I am not sure that it is. I am suggesting that once we get a `lastFoundRowId`, we know we won't find a lower one (assuming that the rowId is always strictly increasing) from the ordinal map, and that means we don't need to read from the map to try to see if there is an exact match for the row id.

Maybe I am missing an edge case in the `DiskBinarySearch` logic though?

> (3) is fine, but then it means we anticipate changes in read patterns, so I'll skip assert and log warning instead.

I don't see any reason to expect the read pattern to change, and the assert will force anyone changing the pattern to know that they are breaking an assumption.